internal/security/fileprovider/provider.go

```diff
@@ -59,6 +69,130 @@ func (p *fileTokenProvider) SetConfiguration(secretConfig secretstoreclient.Secr
+			"edgex-service-name": serviceName,
+		}
+
+		// Always create a policy with this name
+		policyName := "edgex-service-" + serviceName
+
+		policyBytes, err := json.Marshal(servicePolicy)
+		if err != nil {
+			p.logger.Error(fmt.Sprintf("failed encode service policy for %s: %s", serviceName, err.Error()))
+			return err
+		}
+
+		if _, err := p.vaultClient.InstallPolicy(privilegedToken, policyName, string(policyBytes)); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to install policy %s: %s", policyName, err.Error()))
+			return err
+		}
+
+		var createTokenResponse interface{}
+
+		if _, err = p.vaultClient.CreateToken(privilegedToken, createTokenParameters, &createTokenResponse); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to create vault token for service %s: %s", serviceName, err.Error()))
+			return err
+		}
+
+		outputTokenDir := filepath.Join(p.tokenConfig.OutputDir, serviceName)
+		outputTokenFilename := filepath.Join(outputTokenDir, p.tokenConfig.OutputFilename)
+		if err := p.fileOpener.MkdirAll(outputTokenDir, os.FileMode(0700)); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to create base directory path(s) %s: %s", outputTokenDir, err.Error()))
+			return err
+		}
+
+		p.logger.Info(fmt.Sprintf("creating token file %s", outputTokenFilename))
+		writeCloser, err := p.fileOpener.OpenFileWriter(outputTokenFilename, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.FileMode(0600))
+		if err != nil {
+			p.logger.Error(fmt.Sprintf("failed open token file for writing %s: %s", outputTokenFilename, err.Error()))
+			return err
+		}
+		defer func() {
+			if writeCloser != nil {
+				writeCloser.Close()
+			}
+		}()
+
+		permissionable, ok := writeCloser.(permissionable)
+		if ok {
+			if serviceConfig.FilePermissions != nil &&
+				(serviceConfig.FilePermissions).ModeOctal != nil {
+				mode, err := strconv.ParseInt(*(serviceConfig.FilePermissions).ModeOctal, 8, 32)
+				if err != nil {
+					p.logger.Error(fmt.Sprintf("invalid file mode %s: %s", *(serviceConfig.FilePermissions).ModeOctal, err.Error()))
+					return err
+				}
+				if err := permissionable.Chmod(os.FileMode(mode)); err != nil {
+					p.logger.Error(fmt.Sprintf("failed to set file mode on %s: %s", outputTokenFilename, err.Error()))
+					return err
+				}
+			}
+			if serviceConfig.FilePermissions != nil &&
+				(serviceConfig.FilePermissions).Uid != nil &&
+				(serviceConfig.FilePermissions).Gid != nil {
+				err := permissionable.Chown(*(serviceConfig.FilePermissions).Uid, *(serviceConfig.FilePermissions).Gid)
+				if err != nil {
+					p.logger.Error(fmt.Sprintf("failed to set file user/group on %s: %s", outputTokenFilename, err.Error()))
+					return err
+				}
+			}
+		}
+
+		json.NewEncoder(writeCloser).Encode(createTokenResponse) // Write resulting token
+
+		if err := writeCloser.Close(); err != nil {
+			p.logger.Error(fmt.Sprintf("failed to close %s: %s", outputTokenFilename, err.Error()))
+			return err
+		} else {
```

Comment:
FWIW, the `else` isn't necessary.  You could omit it and simply have `writeCloser=nil` as the next statement.

Reply:
Simplified.